presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
         assertEquals(query2.getState(), QUEUED);
 
         query1.complete();
-        root.processQueuedQueries();
+        root.updateGroupsAndProcessQueuedQueries();
         assertEquals(query2.getState(), QUEUED);
 
         root.generateCpuQuota(2);
-        root.processQueuedQueries();
+        root.updateGroupsAndProcessQueuedQueries();
         assertEquals(query2.getState(), RUNNING);
     }
 
+    /**
+     * Test resource group cpu usage update by manually invoking the cpu quota regeneration and queue processing methods
+     * that are invoked periodically by the resource group manager
+     */
+    @Test(timeOut = 10_000)
+    public void testCpuUsageUpdateForRunningQuery()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
+        List<InternalResourceGroup> groups = ImmutableList.of(root, child);
+        groups.forEach(group -> {
+            group.setCpuQuotaGenerationMillisPerSecond((new Duration(1, SECONDS)).toMillis());
+            group.setHardCpuLimit(new Duration(1.5, SECONDS));
+            group.setSoftCpuLimit(new Duration(1.5, SECONDS));
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+        });
+
+        MockResourceUpdatingQueryExecution q1 = new MockResourceUpdatingQueryExecution();
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+
+        // The cached cpu usage will be set to 2s (> 1.5s limit) when updateGroupsAndProcessQueuedQueries is invoked
+        q1.consumeCpuTime(new Duration(2, SECONDS));
+        root.updateGroupsAndProcessQueuedQueries();
+
+        // A newly submitted query q2 will get queued, because the cached usage is greater than the limit
+        MockResourceUpdatingQueryExecution q2 = new MockResourceUpdatingQueryExecution();
+        child.run(q2);
+        assertEquals(q2.getState(), QUEUED);
+
+        // Generating cpu quota before the query finishes. The cached cpu usage will come down to 1s (< 1.5s limit)
+        root.generateCpuQuota(1);
+
+        // An incoming query q3 should not get queued since the cached usage has been updated to 1s < 1.5s. This assertion
+        // is useful for testing cpu update during quota generation, without the invocation of updateGroupsAndProcessQueuedQueries.
+        MockResourceUpdatingQueryExecution q3 = new MockResourceUpdatingQueryExecution();
+        child.run(q3);
+        assertEquals(q3.getState(), RUNNING);
```

Comment:
Yes, I added this test to document this behavior. `refreshAndStartQueries` method is run periodically by the `InternalResourceGroupManager`. It contains invocations of `generateCpuQuota` and `updateResourceUsageAndGetDelta` methods. The case mentioned in the test here only arises if a query comes in between invocations of these two methods. This is possible since  `refreshAndStartQueries` is not synchronized.

Reply:
It's definitely not great behavior and could lead to a lot of user confusion, did your change make this worse? If not, then you can address it in a follow up PR.